lwan: use jemalloc (not just ignored buildInput), but not w/musl

dtzWill opened the pull request:
###### Motivation for this change

jemalloc and musl don't get along.

Replaces #77844 .

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

dtzWill committed:
lwan: no jemalloc w/musl

SuperSandro2000 commented on pkgs/servers/http/lwan/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchFromGitHub, pkgconfig, zlib, cmake, jemalloc }:
+{ stdenv, fetchFromGitHub, pkgconfig, zlib, cmake, enableJemalloc ? !stdenv.hostPlatform.isMusl, jemalloc ? null }:
```
```suggestion
{ stdenv, lib, fetchFromGitHub, pkg-config, zlib, cmake, enableJemalloc ? !stdenv.hostPlatform.isMusl, jemalloc }:
```

SuperSandro2000 commented on pkgs/servers/http/lwan/default.nix:
```diff
@@ -13,7 +13,10 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ cmake pkgconfig ];
 
-  buildInputs = [ jemalloc zlib ];
+  buildInputs = [ zlib ] ++ stdenv.lib.optional enableJemalloc jemalloc;
```
```suggestion
  buildInputs = [ zlib ] ++ lib.optional enableJemalloc jemalloc;
```

SuperSandro2000 commented on pkgs/servers/http/lwan/default.nix:
```diff
@@ -13,7 +13,10 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [ cmake pkgconfig ];
```
```suggestion
  nativeBuildInputs = [ cmake pkg-config ];
```

SuperSandro2000 commented on pkgs/servers/http/lwan/default.nix:
```diff
@@ -13,7 +13,10 @@ stdenv.mkDerivation rec {
+  buildInputs = [ zlib ] ++ stdenv.lib.optional enableJemalloc jemalloc;
+
+  # Note: tcmalloc and mimalloc are also supported (and normal malloc)
+  cmakeFlags = stdenv.lib.optional enableJemalloc "-DUSE_ALTERNATIVE_MALLOC=jemalloc";
```
```suggestion
  cmakeFlags = lib.optional enableJemalloc "-DUSE_ALTERNATIVE_MALLOC=jemalloc";
```

SuperSandro2000 commented on pkgs/servers/http/lwan/default.nix:
```diff
@@ -13,7 +13,10 @@ stdenv.mkDerivation rec {
+  # Note: tcmalloc and mimalloc are also supported (and normal malloc)
+  cmakeFlags = stdenv.lib.optional enableJemalloc "-DUSE_ALTERNATIVE_MALLOC=jemalloc";
 
   meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```

dtzWill committed:
lwan: indicate jemalloc should be used, or it won't be

(cherry picked from commit d45a92fba512e8aeaf7bb904077ca8cd4ad988a8)

dtzWill committed:
lwan: fix build w/musl, disable PIE

dtzWill commented:
Thank you for your feedback, updated this to latest tree and addressed review feedback.

Included another commit to fix build w/musl (not needed previously), disabling PIE in our hardening, not sure if issue with our musl toolchain or what.  lwan itself checks for and uses `-no-pie` while building, if that's relevant.

Without that change, FWIW, it fails with this:

```
/nix/store/cv3vpyz5p0ny2cdcsc65dyz1vwg28nb9-binutils-2.35.2/bin/ld: /nix/store/6jwdsxmvjm4l59q1ia0sy0adppchy3dg-gcc-10.3.0/lib/gcc/x86_64-unknown-linux-musl/10.3.0/crtbegin.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a PIE object
/nix/store/cv3vpyz5p0ny2cdcsc65dyz1vwg28nb9-binutils-2.35.2/bin/ld: /nix/store/6jwdsxmvjm4l59q1ia0sy0adppchy3dg-gcc-10.3.0/lib/gcc/x86_64-unknown-linux-musl/10.3.0/crtend.o: relocation R_X86_64_32 against `.ctors' can not be used when making a PIE object; recompile with -fPIE
```

SuperSandro2000 commented:
@ofborg eval

dtzWill commented:
Thanks!
